core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2555,13 +2555,26 @@ class KafkaApis(val requestChannel: RequestChannel,
 
   def handleAlterReplicaLogDirsRequest(request: RequestChannel.Request): Unit = {
     val alterReplicaDirsRequest = request.body[AlterReplicaLogDirsRequest]
-    val responseMap = {
+    def sendResponseCallback(requestThrottleMs: Int): AlterReplicaLogDirsResponse = {
```

Comment:
I changed it because both branches of the `if authorize() ...` require the `requestThrottleMs` to set on the `AlterReplicaLogDirsResponseData`. I can set the throttle outside the `if` with something like

```scala
sendResponseMaybeThrottle(request, requestThrottleMs => {
      response.data().setThrottleTimeMs(requestThrottleMs)
      response
    })
```

It would then be more natural to call the `alterReplicaDirsRequest.getErrorResponse(Errors.CLUSTER_AUTHORIZATION_FAILED.exception)` which doesn't take a throttle (since we'll be setting it later), but that requires overriding that method so that the return type is not `AbstractResponse`. 

It's a bit unsatisfying, but did you have a better idea?

Reply:
I was thinking about something like this:
```
  def handleAlterReplicaLogDirsRequest(request: RequestChannel.Request): Unit = {
    val alterReplicaDirsRequest = request.body[AlterReplicaLogDirsRequest]
    if (authorize(request.context, ALTER, CLUSTER, CLUSTER_NAME)) {
      val result = replicaManager.alterReplicaLogDirs(alterReplicaDirsRequest.partitionDirs.asScala)
      sendResponseMaybeThrottle(request, requestThrottleMs =>
        new AlterReplicaLogDirsResponse(new AlterReplicaLogDirsResponseData()
          .setResults(result.groupBy(_._1.topic).map {
            case (topic, errors) => new AlterReplicaLogDirsResponseData.AlterReplicaLogDirTopicResult()
              .setTopicName(topic)
              .setPartitions(errors.map { 
                case (tp, error) => new AlterReplicaLogDirsResponseData.AlterReplicaLogDirPartitionResult()
                  .setPartitionIndex(tp.partition)
                  .setErrorCode(error.code)
              }.toList.asJava)
            }.toList.asJava)
          .setThrottleTimeMs(requestThrottleMs))
    } else {
      sendResponseMaybeThrottle(request, requestThrottleMs =>
        alterReplicaDirsRequest.getErrorResponse(requestThrottleMs, Errors.CLUSTER_AUTHORIZATION_FAILED.exception))
    }
  }
```

I find it a bit easier to read and it is also a bit more consistent with the implementation of the other handlers.